README.md

```diff
@@ -26,7 +26,7 @@ If you want to contribute to OpenFisca-Core itself, welcome! To install it local
 ```bash
 git clone https://github.com/openfisca/openfisca-core.git
 cd openfisca-core
-pip install --editable .[dev] --use-deprecated=legacy-resolver
+pip install --editable .[dev]
```

Comment:
I invite you to propose this change in a separate pull request, in order to have a specific place to discuss about the new dependency resolver, and the best strategy to use it —or not.

Spoiler alert: it can be [way too slow](https://github.com/pypa/pip/issues/9187#issuecomment-806905059), which has proven to be painful in time-sensitive contexts as CI and [test deployments](https://app.netlify.com/sites/openfisca-doc/deploys/6060eb0268d09100073b4b5b).

Reply:
I've just opened this pull request to fix this very same issue on the doc https://github.com/openfisca/openfisca-doc/pull/235

Reply:
This very branch [takes 2 seconds](https://app.circleci.com/pipelines/github/openfisca/openfisca-core/231/workflows/b03f2033-6319-49ba-b822-67bdeb6fb5d5/jobs/6780) to install dependencies in CI. Based on the issue you linked to (https://github.com/pypa/pip/issues/9187), I understand that long builds could happen from a specific set of requirements, not randomly. If I understood correctly, is there anything that makes us suspect our set of requirements could generate such long builds? 🙂

Reply:
You're right, pip tries to download all the intervals, and when it comes to NumPy and the like that adds a [lot of overhead](https://app.circleci.com/pipelines/github/openfisca/openfisca-doc/424/workflows/f4145b16-44e2-4d61-8e76-b5adb9f40615/jobs/1193).

The problem is with the first run, then libraries should normally be saved to cache in CircleCI, per branch —one possible workaround being a stricter version interval (like `==`) but aren't we then defeating the purpose of the tool? It looks like next version of pip is going to adresses this backtracking problem.

I know @sarafalamaki and @sandcha run into problems as well with the new resolver —I don't know if the same problem or others.

Reply:
Just in case my point is not to not use the new resolver but to do so in a new pull request so to document and to address the problems users may have encountered that lead to the actual use of the legacy resolver.

Reply:
BTW found this previous discussion, very interesting for reference https://github.com/openfisca/extension-template/issues/10.

Reply:
> This very branch takes 2 seconds to install dependencies in CI.

That build [is using](https://circleci.com/api/v1.1/project/github/openfisca/openfisca-core/6780/output/104/0?file=true&allocation-id=606105d7a1285115cf0d3ccc-0-build%2F5E874C36) the legacy resolver.

---

I've also found the strategies proposed by the Pip folks https://pip.pypa.io/en/latest/user_guide/#dependency-resolution-backtracking, which leds me to think that, before PyPi provides package's metadata as an API, we will have to find a workaround for this problem.

Good news is there are some dependencies we know as a fact can be pinned, like `numpy`, `numexpr`,  `scipy`, and so on, which could be a solution.

I run the following test —I had to cancel at some point:

##### First attempt

As it is now.

```python
general_requirements = [
    'dpath >= 1.5.0, < 2.0.0',
    'pytest >= 4.4.1, < 6.0.0',  # For openfisca test
    'numpy >= 1.11, < 1.19',
    'psutil >= 5.4.7, < 6.0.0',
    'PyYAML >= 3.10',
    'sortedcontainers == 2.2.2',
    'numexpr >= 2.7.0, <= 3.0',
    ]

api_requirements = [
    'werkzeug >= 1.0.0, < 2.0.0',
    'flask == 1.1.2',
    'flask-cors == 3.0.10',
    'gunicorn >= 20.0.0, < 21.0.0',
    ]

dev_requirements = [
    'autopep8 >= 1.4.0, < 1.6.0',
    'flake8 >= 3.9.0, < 4.0.0',
    'flake8-bugbear >= 19.3.0, < 20.0.0',
    'flake8-print >= 3.1.0, < 4.0.0',
    'pytest-cov >= 2.6.1, < 3.0.0',
    'mypy >= 0.701, < 0.800',
    'openfisca-country-template >= 3.10.0, < 4.0.0',
    'openfisca-extension-template >= 1.2.0rc0, < 2.0.0'
    ] + api_requirements
```

```
pip freeze | grep -v "^-e" | xargs pip uninstall -y
pip cache purge
pip install --upgrade pip
time pip install --editable .[dev] --upgrade

...

ERROR: Operation cancelled by user

real	17m37.289s
user	2m36.219s
sys	0m17.993s
```

##### Second attempt

Tightening dependency versions.

```python
general_requirements = [
    'dpath ~= 1.5.0',
    'pytest ~= 5.4.0',  # For openfisca test
    'numpy == 1.18.5',
    'psutil ~= 5.8.0',
    'PyYAML ~= 5.4.0',
    'sortedcontainers == 2.2.2',
    'numexpr ~= 2.7.0',
    ]

api_requirements = [
    'werkzeug ~= 1.0.0',
    'flask == 1.1.2',
    'flask-cors == 3.0.10',
    'gunicorn ~= 20.1.0',
    ]

dev_requirements = [
    'autopep8 ~= 1.5.0',
    'flake8 ~= 3.9.0',
    'flake8-bugbear ~= 19.8.0',
    'flake8-print ~= 3.1.0',
    'pytest-cov ~= 2.11.0',
    'mypy == 0.790',
    'openfisca-country-template >= 3.10.0, < 4.0.0',
    'openfisca-extension-template >= 1.2.0rc0, < 2.0.0',
    ] + api_requirements
```

```
pip freeze | grep -v "^-e" | xargs pip uninstall -y
pip cache purge
pip install --upgrade pip
time pip install --editable .[dev] --upgrade

...

ERROR: Operation cancelled by user

real	7m22.893s
user	6m25.452s
sys	0m2.349s
```

### Third attempt

As it is but pre-calculating some dependencies with `pip-tools`.

It works! I mean, it fails, but the resolver works: we have an incompatible circular dependency.

Even if we reduce the time it take pip to resolve to a couple of minutes, we have core depending on country-template, and country-template depending on core, so it'll inevitably fail.

```
pip freeze | grep -v "^-e" | xargs pip uninstall -y
pip cache purge
pip install --upgrade pip
pip install pip-tools
time ( echo "--editable .[dev]" | pip-compile --output-file=- > /tmp/requirements.txt - -qo- && pip install -r /tmp/requirements.txt --upgrade )

...

Requirement already satisfied: OpenFisca-Core[web-api]<36.0.0,>=35.0.0 in /Users/hyperion/Sites/openfisca/openfisca-core (from openfisca-country-template==3.12.5->-r /tmp/requirements.txt (line 65)) (35.3.0)
Collecting OpenFisca-Core[web-api]<36.0.0,>=35.0.0
  Downloading OpenFisca_Core-35.2.0-py3-none-any.whl (162 kB)
     |████████████████████████████████| 162 kB 1.3 MB/s 
  Downloading OpenFisca_Core-35.1.1-py3-none-any.whl (162 kB)
     |████████████████████████████████| 162 kB 1.0 MB/s 
  Downloading OpenFisca_Core-35.1.0-py3-none-any.whl (162 kB)
     |████████████████████████████████| 162 kB 265 kB/s 
  Downloading OpenFisca_Core-35.0.5-py3-none-any.whl (162 kB)
     |████████████████████████████████| 162 kB 2.1 MB/s 
  Downloading OpenFisca_Core-35.0.4-py3-none-any.whl (162 kB)
     |████████████████████████████████| 162 kB 1.2 MB/s 
  Downloading OpenFisca_Core-35.0.3-py3-none-any.whl (162 kB)
     |████████████████████████████████| 162 kB 2.6 MB/s 
  Downloading OpenFisca_Core-35.0.1-py3-none-any.whl (162 kB)
     |████████████████████████████████| 162 kB 1.9 MB/s 
  Downloading OpenFisca_Core-35.0.0-py3-none-any.whl (162 kB)
     |████████████████████████████████| 162 kB 2.0 MB/s 

...

ERROR: Cannot install openfisca-core 35.3.0 (from /Users/hyperion/Sites/openfisca/openfisca-core), openfisca-core[web-api]==35.0.0, openfisca-core[web-api]==35.0.1, openfisca-core[web-api]==35.0.3, openfisca-core[web-api]==35.0.4, openfisca-core[web-api]==35.0.5, openfisca-core[web-api]==35.1.0, openfisca-core[web-api]==35.1.1, openfisca-core[web-api]==35.2.0 and openfisca-core[web-api]==35.3.0 because these package versions have conflicting dependencies.

The conflict is caused by:
    The user requested openfisca-core 35.3.0 (from /Users/hyperion/Sites/openfisca/openfisca-core)
    openfisca-core[web-api] 35.3.0 depends on openfisca-core 35.3.0 (Installed)


real	0m49.232s
user	0m23.754s
sys	0m1.487s
...

---

Considering the results of those tests, I'm afraid this change break things for users, not just because of the time overhead but as we've just discovered also because of the circular dependencies.

Reply:
> If I understood correctly, is there anything that makes us suspect our set of requirements could generate such long builds? 🙂
> > You're right, pip tries to download all the intervals, and when it comes to NumPy and the like that adds a lot of overhead.

Well... when I pinned NumPy's version I discovered the resolver backtracks following dependencies dependencies, which presents two problems:

1. Backtracking is exponential.
2. We don't have a have any control over the way our dependencies manage their own dependencies.

The following are the [solutions proposed](https://pip.pypa.io/en/latest/user_guide/#possible-ways-to-reduce-backtracking-occurring) by the Pip folks:

> 1. ALLOW PIP TO COMPLETE ITS BACKTRACKING

Already tried, it doesn't work unless we solve our two major problems (wild dependencies' dependencies and circular dependencies).

> 2. REDUCE THE NUMBER OF VERSIONS PIP WILL TRY TO BACKTRACK THROUGH

Also already tried, cf. supra.

> 3. USE CONSTRAINT FILES OR LOCKFILES

This could be easily done with `pip-tools`, `pipgrip`, `Pipenv`, `Poetry`, etc.

We'd still have to solve the circular dependency problem, though, unless we ignore them —given `pip` allows us to do so, or using `pipgrip` as an alternative.

> 4. BE MORE STRICT ON PACKAGE DEPENDENCIES DURING DEVELOPMENT

Same as 1. and 2.

😃

Reply:
Thanks a lot for this thorough and documented exploration of the problem space @maukoquiroga! 🙏😃